src/astro/databases/google/bigquery.py

```diff
@@ -135,3 +146,96 @@ def merge_table(
             )
             statement += f" WHEN MATCHED THEN {update_statement}"
         self.run_sql(sql_statement=statement)
+
+    def check_optimised_path_and_transfer(
+        self,
+        source_file: File,
+        target_table: Table,
+        chunk_size: int = DEFAULT_CHUNK_SIZE,
+        if_exists: LoadExistStrategy = "replace",
+        **kwargs,
+    ) -> bool:
+        """
+        Checks if optimised path for transfer between File location to database exists
+        and if it does, it transfers it and returns true else false.
+        """
+        method_name = self.OPTIMIZED_PATHS.get(source_file.location.location_type)
+        if method_name:
+            transfer_method = self.__getattribute__(method_name)
+            transfer_method(
+                source_file=source_file,
+                target_table=target_table,
+                chunk_size=chunk_size,
+                if_exists=if_exists,
+                **kwargs,
+            )
+            return True
+
+        return False
+
+    @staticmethod
+    def get_project_id_from_conn(conn: connection) -> str:
+        """
+        Get project id from conn
+        :param conn: Airflow's connection
+        """
+        if conn.extra and conn.extra.get("project"):
+            return str(conn.extra.get("project"))
```

Comment:
```suggestion
            return conn.extra["project"]
```

This might fix it and not need `str()`

Reply:
We already check on L182 that "project" key is present in extra.

Also you might actually want `extra_dejson` not `extra`

Reply:
Resolved.